colcon_sanitizer_reports/sanitizer_log_parser.py

```diff
@@ -0,0 +1,195 @@
+            )
+        )
+
+    @staticmethod
+    def _make_lock_order_inversion_keys(*, lines: Tuple[str]) -> Tuple[str]:
+        # There are two stack traces involved in a "lock-order-inversion" error, both in the same
+        # subsection and with identical header lines.
+        return _KeysFactory._make_keys_base(
+            lines=lines,
+            stack_begin_regexes=(
+                r'^\s+Mutex M\d+ acquired here while holding mutex M\d+ in .*$',
+                r'^\s+Mutex M\d+ acquired here while holding mutex M\d+ in .*$'
+            )
+        )
+
+    @staticmethod
+    def make_keys(*, lines: Tuple[str], error_name: str) -> Tuple[str]:
+        if error_name == 'data race':
+            return _KeysFactory._make_data_race_keys(lines=lines)
+        elif error_name == 'detected memory leaks':
+            return _KeysFactory._make_detected_memory_leaks_keys(lines=lines)
+        elif error_name == 'lock-order-inversion':
+            return _KeysFactory._make_lock_order_inversion_keys(lines=lines)
+        else:
+            return _KeysFactory._make_default_keys(lines=lines)
+
+
+class _SubSection:
+    keys: Tuple[str]
+
+    def __init__(self, *, lines: Tuple[str], error_name: str) -> None:
+        self.keys = _KeysFactory.make_keys(lines=lines, error_name=error_name)
+
+
+class _Section:
+    """Contents for a single Sanitizer output section."""
+
+    error_name: str
+    sub_sections: Tuple[_SubSection]
+
+    def __init__(self, *, lines: Tuple[str]) -> None:
+        # Section name comes after 'Sanitizer: ', and before any open paren or hex number.
+        self.error_name = re.match(
+            r'^.*Sanitizer: (?P<error_name>.+?)( \(| 0x[\dabcdef]+|\s*$)', lines[0]
+        ).groupdict()['error_name']
+
+        # Divide into _SubSections. SubSections begin with a line that is not indented.
+        sub_section_lines: List[str] = []
+        sub_sections: List[_SubSection] = []
+        for line in lines:
+            if re.match(r'^\S.*$', line) is not None and sub_section_lines:
+                sub_sections.append(
+                    _SubSection(lines=tuple(sub_section_lines), error_name=self.error_name)
+                )
+                sub_section_lines = []
+
+            sub_section_lines.append(line)
+        else:
+            sub_sections.append(
+                _SubSection(lines=tuple(sub_section_lines), error_name=self.error_name)
+            )
+
+        self.sub_sections = tuple(sub_sections)
+
+
+class SanitizerLogParser:
+    """Parses logged sanitizer errors and warnings and generates report."""
+
+    def __init__(self) -> None:
+        # Holds count of errors seen. This is what will be in the report.
+        self._counts = defaultdict(int)
+
+        # Current package and sections that are partially parsed
+        self._package: Optional[str] = None
+        self._section_lines_by_prefix: Dict[str: List[str]] = {}
+
+    @property
+    def csv(self) -> str:
```

Comment:
Having a property do something which is not `O(1)` is a tad dangerous. I feel that `build_csv` is closer to the truth.